### ISSUE TITLE
API for constructing stateless ViewStore

### ISSUE THREAD

**Issue by LK** [pointfreeco/swift-composable-architecture#43]
I find that I occasionally need to use `WithViewStore` only for the purpose of being able to send actions, without ever accessing any state on the view store directly (generally this happens when I introduce some more local store somewhere down the tree, with a `ForEachStore` or an `IfLetStore`). For example, if I'm rendering a table section with some list elements, I might do something like this:

```swift
WithViewStore(self.store) { viewStore in
    Section(header: Text("Header Text")) {
    ForEachStore(self.store.scope(state: \.books.elements, action: ReadingListAction.rowInteraction)) { bookStore in
            WithViewStore(bookStore) { bookViewStore in
                BookRow(title: bookViewStore.title)
                    .onAppear { viewStore.send(.rowAppeared(bookViewStore.id)) }
                    .onDisappear { viewStore.send(.rowDisappeared(bookViewStore.id)) }
            }
        }
    }
}
```

In this case, I want the `rowAppeared` and `rowDisappeared` actions to be sent to the top-level reducer, so I need the global `viewStore`, however I never need to touch any of its state directly.

This works fine, however I'm now re-rendering the entire section every time any of my rows are updated, which is unnecessary. What I actually want is to have the top-level `viewStore` ignore any state changes and just provide an outlet to send actions.

In these cases, it seems like the correct thing to do is to `scope` the store down to an empty state struct. Doing so is a bit cumbersome, so I've been using a custom helper:

```swift
public extension Store {
    struct Stateless: Equatable {}
    func stateless() -> Store<Stateless, Action> {
        return self.scope(state: { _ in Stateless() })
    }
}
```

There's a handful of different ways this could be expressed in the API (a `WithStatelessViewStore` helper might be better, because I see no other purpose for a stateless store other than to feed a `ViewStore`). Maybe the problem I'm trying to solve is a code smell and I should just refactor my actions or views to avoid the need for this altogether; either way, curious to hear thoughts.

**Comment by Odrakir** [comment]
`WithViewStore` can take a closure to remove duplicates, you could maybe make it so it always return true? Even creating a `WithStatelessViewStore` wrapper that does this for you. 

This is a hacky solution, so don’t take it too seriously, I just wanted to share my two cents.

**Comment by stephencelis** [comment]
@LK Thanks for starting this discussion! This is definitely a common enough problem that it would be nice to have a dedicated helper to solve it. We just haven't quite figured out the name that feels "right" yet.

@Odrakir Yup! You can create a stateless store by erasing its state to `Void`:

```swift
store.scope(state: { _ in () }, removeDuplicates: ==) { … }

// When "SE-0283: Tuples conform to Equatable, Comparable, and Hashable" lands:
store.scope(state: { _ in () }) { … } 
```

Neither of these solutions are great, though, so we're down to workshop things a bit.

**Comment by stephencelis** [comment]
Some designs @mbrandonw and I have previously discussed:

1. Introduce module-level `noState` and `noAction` transform functions:

    ```swift
    let sendOnlyStore = store.scope(state: noState)
    let readOnlyStore = store.scope(action: noAction)
    ```

    Pro: reuse existing methods.
    Con: introduce new module-level functions, which are slightly less discoverable and can feel "wrong" to some folks.

2. Introduce methods that bundle up this behavior:

    ```swift
    let sendOnlyStore = store.withoutState()
    // or `.withoutState(action: { .localAction($0) })`
    let readOnlyStore = store.withoutActions()
    // or `.withoutActions(state: { $0.localState })`
    ```

   Are there better names? `ignoring{State,Action}`?

**Comment by kaishin** [comment]
I would definitely vote for 2. The first alternative doesn't read particularly well, even when putting aside other concerns.

As for the name, `ignoringX` could definitely work. Or to match `reducer.optional`, this could be a computed variable named something along the lines of `store.stateless` and `store.actionless` (just spitballing here).

**Comment by stephencelis** [comment]
I've opened https://github.com/pointfreeco/swift-composable-architecture/pull/45 to play around with the property-based API and it's feeling nice! We'll leave discussion open around naming for a bit, but I think `stateless` and `actionless` are definitely top contenders!

**Comment by LK** [comment]
Thanks for the speedy PR @stephencelis! Having thought about it a bit more, I'm starting to like `ignoringState()` or something along those lines more, because a stateless store seems like a bit of an oxymoron — what good is a store without any state, what we're really doing is just ignoring the state. I would agree that a global transform function seems unnecessary here, so I like the way you have the API structured in #45.

I also like the `WithViewStore` extension, seems cleaner than what how I had originally implemented it.

**Comment by LK** [comment]
Implemented in #45
